### ISSUE TITLE
Support YAML (YML) officially 

### ISSUE THREAD

**Issue by iocapps** [Kotlin/kotlinx.serialization#1836]
YAML is heavily used as a config format and far more relevant than some of the other formats that have first class support. I believe YAML should be supported officially instead of relying on 3rd party libraries mentioned in the docs.

**Comment by sandwwraith** [comment]
Non-JSON formats in the kotlinx.serialization have 'experimental' status, so not sure if this can be qualified as first-class support. Besides, what's wrong with 3rd party providers?

**Comment by iocapps** [comment]
> Non-JSON formats in the kotlinx.serialization have 'experimental' status, so not sure if this can be qualified as first-class support. Besides, what's wrong with 3rd party providers?

True, they are experimental, but they are still prioritized over YAML. 
It would be great to have official YAML support from Kotlin team since there is smaller chance of it being abandoned, typically better code quality due to stricter standards, quicker to fix vulnerabilities, etc compared to 3rd party libraries.

**Comment by wakaztahir** [comment]
No one offers kotlin multiplatform yaml serialization , only jvm is supported mostly and others who do support kotlin multiplatform aren't feature complete

**Comment by AngeloFilaseta** [comment]
There are some third-party multiplatform libraries available that enable the YAML format.  [kaml](https://github.com/charleskorn/kaml)  and [yamlkt](https://github.com/Him188/yamlkt) are two examples.

I strongly agree with @iocapps . Another thing to point out is that third-party libraries do not always offer the same level of consistency.
As an example, I discovered that the yamlkt library exhibits distinct behavior and results compared to the JSON serializer provided by kotlinx.serialization, particularly in scenarios involving polymorphic serialization.
I would love to witness the YAML format's support within kotlinx.serialization, even if in an experimental stage.

**Comment by dxinli** [comment]
how apply,is support?kotlin use yaml for multiplatform is so bad,kotlinx is beatiful , support yaml would amazing

**Comment by sschuberth** [comment]
> [kaml](https://github.com/charleskorn/kaml) and [yamlkt](https://github.com/Him188/yamlkt) are two examples.

Note that by now, the "kaml" project has been archived, and "yamlkt" didn't see any updates in about 10 months.

**Comment by AngeloFilaseta** [comment]
I think we can all agree that other libraries dropping support might actually be a good opportunity for kotlinx.serialization.

Right now our team is using a custom StringFormat that leverages another YAML library under the hood to convert the JSON produced by Kotlinx.serialization. It works, but it sucks for many uses. Moreover, this implementation of ours is JVM-only (like most libraries).

Having mpp support for YAML directly in the library would be a huge win.

**Comment by hfhbd** [comment]
My use-case:
I have an OpenAPI generator [kfx](https://github.com/hfhbd/kfx), but I currently only support JSON files because I need to use the JsonDecoder to support `x-` extensions. 

While JSON is a subset of YAML, so each Json document is a valid yaml document, the official JsonDecoder cannot/does not extend a (nonexistent/3rd party) YamlDecoder. Also, all 3rd party formats do not support a mapping of a YamlElement to a JsonElement.

As a library author who wants to support Json and YAML, you need to duplicate the custom decoding logic for Json and YAML, regardless their compatibility.

**Comment by sschuberth** [comment]
FYI, there's some interesting related work happening in https://github.com/juanchurtado1991/ghost-serializer/pull/3.
